File: Java/DialogueRuntime.java
import java.time.Duration;
import java.util.List;

public final class DialogueRuntime {
    private final DialogueRenderer renderer;

    public DialogueRuntime(DialogueRenderer renderer) {
        this.renderer = renderer;
    }

    public DialogueResult execute(Dialogue dialogue) {
        DialogueResult result = new DialogueResult();

        runSteps(dialogue.steps(), result);

        result.setCompleted(true);
        dialogue.triggerEnd();

        return result;
    }

    private void runSteps(List<DialogueSteps.IStep> steps, DialogueResult result) {
        for (DialogueSteps.IStep step : steps) if (!runStep(step, result)) return;
    }

    private boolean runStep(DialogueSteps.IStep step, DialogueResult result) {
        if (step instanceof DialogueSteps.Say(String text)) {
            renderer.renderText(text);
            return true;
        }

        if (step instanceof DialogueSteps.Pause(Duration duration)) {
            renderer.wait(duration);
            return true;
        }

        if (step instanceof DialogueSteps.End) return false;

        if (step instanceof DialogueSteps.Choice choice) {
            runChoice(choice, result);
            return true;
        }

        return true;
    }

    private void runChoice(DialogueSteps.Choice choice, DialogueResult result) {
        List<DialogueOption> available = choice.options().stream().filter(DialogueOption::isAvailable).toList();
        if (available.isEmpty()) return;

        renderer.renderChoice(choice.prompt(), available.stream().map(DialogueOption::label).toList());

        DialogueOption selected = available.get(renderer.getChoiceInput(available.size()));

        result.recordChoice(choice.prompt(), selected.label());

        if (selected.branch() != null) runSteps(selected.branch().steps(), result);
    }
}

File: C#/DialogueRuntime.cs
namespace Dialful;

public sealed class DialogueRuntime(IDialogueRenderer renderer)
{
    private readonly IDialogueRenderer Renderer = renderer;

    public DialogueResult Execute(Dialogue dialogue)
    {
        DialogueResult result = new DialogueResult();

        Run(dialogue.Steps, result);
        result.Completed = true;
        dialogue.End();

        return result;
    }

    private bool Run(IReadOnlyList<IDialogueStep> steps, DialogueResult result)
    {
        foreach (IDialogueStep step in steps) if (!Step(step, result)) return false;

        return true;
    }

    private bool Step(IDialogueStep step, DialogueResult result)
    {
        switch (step)
        {
            case SayStep say:
                Renderer.RenderText(say.Text);
                return true;

            case PauseStep pause:
                Renderer.Wait(pause.Duration);
                return true;

            case EndStep:
                return false;

            case ChoiceStep choice:
                Choice(choice.Prompt, choice.Options.ToList(), result);
                return true;

            default:
                return true; // skip unknown types
        }
    }

    private void Choice(string prompt, List<DialogueOption> options, DialogueResult result)
    {
        var available = options.Where(o => o.Available).ToList();
        if (available.Count == 0) return; // skip if no options available

        Renderer.RenderChoice(prompt, available.Select(o => o.Label).ToList());

        DialogueOption selected = available[Renderer.GetChoiceInput(available.Count)];

        result.RecordChoice(prompt, selected.Label);
        if (selected.Branch != null) Run(selected.Branch.Steps, result);
    }
}
